raiden/transfer/mediated_transfer/mediator.py

```diff
@@ -840,7 +843,7 @@ def events_for_onchain_secretreveal_if_closed(
         transfers_pair: List[MediationPairState],
         secret: Secret,
         secrethash: SecretHash,
-) -> Sequence[ContractSendSecretReveal]:
+) -> List[Event]:
```

Comment:
But the type is more accurate as a Sequence (aka list) of `ContractSendSecretReveal` since we are sure from the code that only this event is returned. There was a whole conversation about it when Yoichi made this change here: https://github.com/raiden-network/raiden/pull/3219#discussion_r245232593

Reply:
Shouldn't mypy complain about this?

Reply:
> But the type is more accurate as a Sequence (aka list) of `ContractSendSecretReveal` since we are sure from the code that only this event is returned

Using a more specific type misses the point. These functions that start with a prefix `events_for` are returning a `List[Event]` because that is the result of applying a state change. The number of events, their types, their order, must not be of any importance to the rest of the code base, therefore the type here should be erased.

Reply:
> The number of events, their types, their order, must not be of any importance to the rest of the code base, therefore the type here should be erased

To me types are also the best type of documentation. I just see the signature and I know immediately what this function returns. So the more specific it is the better.

I think your point is that these `events_for` always have the same return type as an interface so you want the type to be generic?

Reply:
My point is that type erasure is the goal.

Your goal and mine are at odds. You want to have the explicit type there as a form of documentation for a human, I want to *not* have the type in there because I don't want a human to not mess up with it (same reason why we would have a private function in a class).

Reply:
We can do two things:

- Move the documentation about the event type from the type signature to the function docstring.
- Move the place where type erasure takes place.

I'm fine with either, but 2 needs a bit of thought on the boundaries of the code

Reply:
I'm merging this to avoid conflicts, since this is a one-liner if you want it's easy to fix in a follow up PR.

Reply:
> My point is that type erasure is the goal.

Okay. So essentially the generic interface comment I made above then. And you want concrete types to be erased and expose with generics through this interface.

I explained why I like the concrete type there. You also explained why you want type erasure but I have difficulty understanding your argument so I would appreciate some help. You wrote:

> I want to _not_ have the type in there because I don't want a human to not mess up with it (same reason why we would have a private function in a class).

So you want to prevent human mistakes, and you bring the private function in a class as an example which we have to prevent humans from exposing a private function to the outside. 

What I don't understand is how does the generic type interface help to prevent human mistakes here?

Reply:
> What I don't understand is how does the generic type interface help to prevent human mistakes here?

Because:

> The number of events, their types, their order, must not be of any importance to the rest of the code base, therefore the type here should be erased.

These list of events are not of a specific type, they are a list of events:

https://github.com/raiden-network/raiden/blob/bdfb4c8c43aa8fe3c326734a821ef4013ee78066/raiden/transfer/architecture.py#L290-L302

And this is used through out the code:

https://github.com/raiden-network/raiden/blob/bdfb4c8c43aa8fe3c326734a821ef4013ee78066/raiden/transfer/mediated_transfer/initiator.py#L82

https://github.com/raiden-network/raiden/blob/bdfb4c8c43aa8fe3c326734a821ef4013ee78066/raiden/transfer/mediated_transfer/initiator.py#L162

https://github.com/raiden-network/raiden/blob/master/raiden/transfer/mediated_transfer/initiator_manager.py#L109-L110

etc

Reply:
Thank your for the explanation.

> These list of events are not of a specific type, they are a list of events:

All right but my point stands that the particular function returns always one type of event and I think it's nice for documentation to reflect that in the types.

I still don't see how this approach prevents human errors.

Perhaps there could be another point where the type erasure happens so that both of our requirements are satisfied.

---

That said I am not going to continue this conversation as I think I understood your point but don't 100% agree and I think from here and on we could turn this easily into bike shedding argument.

I can live with the change.